File: 1026. Maximum-Difference-Between-Node-and-Ancestor/Solution1026.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
 var maxAncestorDiff = function(root) {
    return helper(root, root.val, root.val);
    function helper(root, minVal, maxVal) {
        if (root == null) {
            return maxVal - minVal;
        }
        minVal = Math.min(minVal, root.val);
        maxVal = Math.max(maxVal, root.val);
        let left = helper(root.left, minVal, maxVal);
        let right = helper(root.right, minVal, maxVal);
        return Math.max(left, right);
    }
};

File: 1026. Maximum-Difference-Between-Node-and-Ancestor/Solution1026.go
import "math"

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func maxAncestorDiff(root *TreeNode) int {
	return helper(root, root.Val, root.Val)
}

func helper(root *TreeNode, minVal int, maxVal int) int {
	if root == nil {
		return maxVal - minVal
	}
	var left, right int
	minVal = int(math.Min(float64(minVal), float64(root.Val)))
	maxVal = int(math.Max(float64(maxVal), float64(root.Val)))
	left = helper(root.Left, minVal, maxVal)
	right = helper(root.Right, minVal, maxVal)
	return int(math.Max(float64(left), float64(right)))
}